caucsejunseo/큐/2025.04.02.요세푸스 문제.c

```diff
@@ -0,0 +1,68 @@
+#define _CRT_SECURE_NO_WARNINGS
+#include<stdio.h> 
+#include<stdlib.h> //malloc�� ����ϱ� ����
+
+int main()
+{
+
+	int i = 0;
+	int N = 0;
+	int K = 0;
+	int count = 0;//���� ��� ��
+	int step = 0;// ���� �� Ȯ��
+	int gun = 0;
+	scanf("%d", &N);
+	scanf("%d", &K);
+
+
+	int* table = (int*)malloc(N * sizeof(int)); //c������ �����Ϳ� �迭�� ����.
+
+	if (table == NULL) {
+		printf("�޸� �Ҵ� ����\n");
+		return 1;
+	}
+	if (table != NULL)
+	{
```

Comment:
이미 첫 번째 if문에서 `table == NULL` 검사를 하기 때문에, 굳이 아래에서 `table != NULL`검사를 할 필요가 없어 보이네요,
불필요한 코드를 제거할 수 있을 것 같습니다!
```suggestion
	if (table == NULL) {
		printf("�޸� �Ҵ� ����\n");
		return 1;
	}
```

Reply:
동적 할당을 하는 이유!
(제가 잘못 알고 있을 수도 있다는 점을 미리 말씀드립니당 ㅎㅎ)

처음 malloc 사용하는 것을 봤을 때 그냥 배열 크기를 나중에 바꿔주면 되는 거 아닌가 생각했었는데 그렇게 하면 메모리가 낭비된다고 하더라구요

예를 들어 arr[5]를 선언하고 크기가 부족해서 나중에 arr[10]으로 다시 선언해서 사용한다면 쓰레기 값이 발생하기 때문에, 
malloc으로 선언해서 나중에 필요할 때  realloc으로 바꾸는 게 좋다고 들었습니다!
![image](https://github.com/user-attachments/assets/9971804b-67be-4349-bf64-9a779e4c5379)

java에서는 자동으로 쓰레기 값을 처리해준다고 하더라구요

Reply:
오...잘 이해하신 것 같아요! 저도 잘 모르는 부분이라 한번 질문드렸습니다.
크기가 다른 배열을 새로이 선언하는 것은 그냥 기존의 배열과는 상관없는 새 배열이 생성되기 때문에 기존 데이터를 유지하지 못하기 때문에 동적 배열을 사용한다고 알고있었는데 일치하는 것 같네요. :-)
이외에도 다른 다양한 이유가 있겠지만 겉핥기로 코멘트를 적기에는 부끄러우니 자료구조를 좀 더 공부해오겠습니다..!